sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala

```diff
@@ -1114,7 +1114,7 @@ class SessionCatalog(
       }
       externalCatalog.dropFunction(db, name.funcName)
     } else if (!ignoreIfNotExists) {
-      throw new NoSuchFunctionException(db = db, func = identifier.toString)
+      throw new NoSuchPermanentFunctionException(db = db, func = identifier.toString)
```

Comment:
Is there no place having the same issue? e.g., `alterFunction`

Reply:
**SPARK-14384** has removed the support of alter function., Later on alter function was added as part of **create or replace** function support (**SPARK-20383**). So when we use create or replace,  the same issue will not come as it creates a new a function if function dies not exist.

Reply:
But looks like `alterFunction` still tried to throw `NoSuchFunctionException`? You mean that can't be reached, right?

Reply:
yes.. that cant be reached.

Reply:
cc @maropu

Reply:
Can we just fix `alterFunction` to match? I don't think it's good idea not to fix because it's unreachable. Catalog itself should function properly alone.

Reply:
Ur, I think so.. +1 to fix `alterFunction`, too.

Reply:
Yes, it makes sense to me.

Reply:
cc @maropu @HyukjinKwon @viirya  I have changed the exception in alter function .